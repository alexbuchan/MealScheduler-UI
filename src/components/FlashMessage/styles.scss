.flash-message {
  bottom: 0px;
  position: absolute;
  width: 100%;

  .alert {
    margin: 0;
    border: none;
    border-radius: 0;
    color: white;
    padding: 20px;
    z-index: 1111;
    display: flex;
    justify-content: space-between;
  }

  .alert p {
    margin: 0;
  }

  .alert-error {
    background-color: #fb2a2a;
  }

  .alert-success {
    background: green;
  }

  .close {
    color: white;
    cursor: pointer;
    margin-right: 10px;
  }
}