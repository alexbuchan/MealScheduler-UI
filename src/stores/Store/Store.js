import { EventEmitter } from 'events';
import Dispatcher from '../../dispatcher/dispatcher';

const CHANGE = 'CHANGE';

class Store extends EventEmitter {
  constructor() {
    super();
    Dispatcher.register(this._registerToActions.bind(this));
  }

  _registerToActions() {
    return 'Not implemented';
  }

  emitChange = () => {
    this.emit(CHANGE);
  }

  addChangeListener(callback) {
    this.on(CHANGE, callback);
  }

  removeChangeListener(callback) {
    this.removeListener(CHANGE, callback);
  }
}

export default Store;