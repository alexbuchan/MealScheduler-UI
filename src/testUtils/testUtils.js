import React from 'react';
import { shallow } from 'enzyme';

export const componentSetup = (Component, props = {}, options = {}) => {
  const instance = shallow(<Component { ...props } />, options);
  return instance;
}

export const findByTestAttribute = (instance, attribute) => {
  const wrapper = instance.find(`[data-test="${attribute}"]`);
  return wrapper;
}