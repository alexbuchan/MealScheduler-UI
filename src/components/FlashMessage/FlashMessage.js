import React from 'react';
import Actions from '../../actions/FlashMessageActions/FlashMessageActions';
import Store from '../../stores/FlashMessageStore/FlashMessageStore';

class FlashMessage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      message: Store.getMessage(),
      open: Store.getOpen(),
      type: Store.getType(),
      duration: this.props.duration || Store.getDuration()
    }
  }

  _onChange = () => {
    this.setState({
      message: Store.getMessage(),
      open: Store.getOpen(),
      type: Store.getType(),
      duration: this.props.duration || Store.getDuration()
    });
  }

  componentDidMount() {
    Store.addChangeListener(this._onChange);
  }

  componentWillUnmount() {
    Store.removeChangeListener(this._onChange);
  }

  handleCloseMessageClick() {
    Actions.closeFlashMessage();
  }

  renderFlashMessage() {
    if (this.state.open) {
      setTimeout(() => Actions.closeFlashMessage(), this.state.duration);

      return (
        <div className={`alert alert-${this.state.type}`}>
          <p></p>
          { this.renderErrorMessage() }
          <span data-test='close-flash-message' className="close" onClick={ () => this.handleCloseMessageClick() }><strong>X</strong></span>
        </div>
      );
    }
    return null;
  }

  renderErrorMessage() {
    if (Object.keys(this.state.message[0]).includes('field')) {
      return (
        <ul>
          { this.state.message.map((err, i) => <li key={i}>{ err.field }: { err.message }</li>) }
        </ul>
      );
    }

    return <li>Error: { this.state.message }</li>
  }

  render() {
    return(
      <div className='flash-message'>
        { this.renderFlashMessage() }
      </div>
    );
  }
}

export default FlashMessage;