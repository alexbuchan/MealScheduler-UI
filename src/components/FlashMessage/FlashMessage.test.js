import FlashMessage from './FlashMessage';
import Actions from '../../actions/FlashMessageActions/FlashMessageActions';
import Store from '../../stores/FlashMessageStore/FlashMessageStore';
import { componentSetup, findByTestAttribute } from '../../testUtils/testUtils';

jest.mock('../../actions/FlashMessageActions/FlashMessageActions');

describe('FlashMessage', () => {
  let instance;

  beforeEach(() => {
    jest.useFakeTimers();
    instance = componentSetup(FlashMessage);
  });

  afterEach(() => {
    jest.clearAllTimers();
    Actions.closeFlashMessage.mockClear();
  });

  describe('When there are no messages in the store', () => {
    it('Should not render', () => {
      expect(instance).toMatchSnapshot();
    });
  });

  describe('When there is a message in the store', () => {
    it('Should render, wait for 3 seconds, then stop rendering', () => {
      // Simulate state change triggered by store
      const error = { data: { error: 'Invalid email' } };
      Store.addErrorMessage(error);

      // component is rendering
      expect(instance).toMatchSnapshot();
      expect(Actions.closeFlashMessage).not.toBeCalled();

      jest.advanceTimersByTime(3000);

      expect(Actions.closeFlashMessage).toHaveBeenCalled();
      Store.closeFlashMessage();
      
      expect(instance).toMatchSnapshot();
    });
  });

  describe('when there are multiple messages in the store', () => {
    it('Should render, wait for 3 seconds, then stop rendering', () => {
      // Simulate state change triggered by store
      const error = { data: { error: ['Invalid email', 'invalid username'] } };
      Store.addErrorMessage(error);

      // component is rendering
      expect(instance).toMatchSnapshot();
      expect(Actions.closeFlashMessage).not.toBeCalled();

      jest.advanceTimersByTime(3000);
      expect(Actions.closeFlashMessage).toHaveBeenCalled();
      Store.closeFlashMessage();
      
      expect(instance).toMatchSnapshot();
    });
  });

  describe('#_onClick', () => {
    it('triggers the closeFlashMessage action when user clicks to close FlashMessage', () => {
      // Simulate state change triggered by store
      const error = { data: { error: ['Invalid email', 'invalid username'] } };
      Store.addErrorMessage(error);

      // Find close button and simulate click
      const closeFlashMessageButton = findByTestAttribute(instance, 'close-flash-message');
      closeFlashMessageButton.simulate('click', { button: 0 });

      expect(Actions.closeFlashMessage).toHaveBeenCalled();
    });
  });

  describe('#renderMessage', () => {
    describe('When error message array length larger than 1', () => {
      it('Should render a list element for each error in the flash message', () => {
        // Simulate state change triggered by store
        const error = {
          data: {
            error: [
              'Invalid email',
              'Username too short'
            ] 
          } 
        };

        Store.addErrorMessage(error);

        // component is rendering
        expect(instance).toMatchSnapshot();
      });
    });

    describe('When error message contains only a string', () => {
      it('Should render only one list item for the error in the flash message', () => {
        const error = { data: { error: ['Invalid email'] } };
        Store.addErrorMessage(error);

        // component is rendering
        expect(instance).toMatchSnapshot();
      });
    });
  });

  describe('#componentWillUnmount', () => {
    it('Should remove the Store event listener', () => {
      const removeChangeListenerSpy = jest.spyOn(Store, 'removeChangeListener');
      instance.unmount();
      expect(removeChangeListenerSpy).toHaveBeenCalled();
    });
  });
});