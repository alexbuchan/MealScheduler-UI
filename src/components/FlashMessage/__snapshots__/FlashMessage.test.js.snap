// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlashMessage #renderErrorMessage When error message array length larger than 1 Should render a list element for each error in the flash message 1`] = `
<div
  className="flash-message"
>
  <div
    className="alert alert-error"
  >
    <p />
    <ul>
      <li
        key="0"
      >
        Invalid email
      </li>
      <li
        key="1"
      >
        Username too short
      </li>
    </ul>
    <span
      className="close"
      data-test="close-flash-message"
      onClick={[Function]}
    >
      <strong>
        X
      </strong>
    </span>
  </div>
</div>
`;

exports[`FlashMessage #renderErrorMessage When error message contains only a string Should render only one list item for the error in the flash message 1`] = `
<div
  className="flash-message"
>
  <div
    className="alert alert-error"
  >
    <p />
    <li>
      Invalid email
    </li>
    <span
      className="close"
      data-test="close-flash-message"
      onClick={[Function]}
    >
      <strong>
        X
      </strong>
    </span>
  </div>
</div>
`;

exports[`FlashMessage When there are no errors in the store Should not render 1`] = `
<div
  className="flash-message"
/>
`;

exports[`FlashMessage When there is an error in the store Should render, wait for 3 seconds, then stop rendering 1`] = `
<div
  className="flash-message"
>
  <div
    className="alert alert-error"
  >
    <p />
    <li>
      Invalid email
    </li>
    <span
      className="close"
      data-test="close-flash-message"
      onClick={[Function]}
    >
      <strong>
        X
      </strong>
    </span>
  </div>
</div>
`;

exports[`FlashMessage When there is an error in the store Should render, wait for 3 seconds, then stop rendering 2`] = `
<div
  className="flash-message"
/>
`;

exports[`FlashMessage when there are multiple error messages in the store Should render, wait for 3 seconds, then stop rendering 1`] = `
<div
  className="flash-message"
>
  <div
    className="alert alert-error"
  >
    <p />
    <ul>
      <li
        key="0"
      >
        Invalid email
      </li>
      <li
        key="1"
      >
        invalid username
      </li>
    </ul>
    <span
      className="close"
      data-test="close-flash-message"
      onClick={[Function]}
    >
      <strong>
        X
      </strong>
    </span>
  </div>
</div>
`;

exports[`FlashMessage when there are multiple error messages in the store Should render, wait for 3 seconds, then stop rendering 2`] = `
<div
  className="flash-message"
/>
`;
